flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/data/util/DataFormatConverters.java

```diff
@@ -1515,21 +1516,40 @@ public CaseClassConverter(TupleTypeInfoBase t, DataType[] fieldTypes) {
         }
 
         @Override
-        RowData toInternalImpl(Product value) {
+        RowData toInternalImpl(Object value) {
             GenericRowData genericRow = new GenericRowData(t.getArity());
             for (int i = 0; i < t.getArity(); i++) {
-                genericRow.setField(i, converters[i].toInternal(value.productElement(i)));
+                try {
+                    genericRow.setField(
+                            i,
+                            converters[i].toInternal(
+                                    PRODUCT_CLASS
```

Comment:
can we store the method in a transient variable that will be filled on access?
It would make the code a bit more performant.
```
PRODUCT_CLASS.getMethod("productElement", int.class)
```

Reply:
Perhaps just a static method?

Reply:
check it out now